File: src/main/java/ch/vilki/jfxldap/backend/CompResult.java
package ch.vilki.jfxldap.backend;

import com.unboundid.ldap.sdk.*;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class CompResult implements java.io.Serializable, Comparable<CompResult> {
    public static enum ENTRY_TYPE {SOURCE, TARGET, BOTH};

    public static enum COMPARE_RESULT {
        ENTRY_EQUAL,
        ENTRY_NOT_EQUAL,
        ONLY_IN_SOURCE,
        ONLY_IN_TARGET,
        ENTRY_EQUAL_BUT_CHILDREN_NOT,
        COMPARE_ERROR,

    };

    private COMPARE_RESULT _compare_result = null;
    private String _dn = null;
    private String _parent_DN = null;
    private ENTRY_TYPE _entry_type;
    private boolean _entryEqual = false;
    private String _rdnDisplay = null;
    private List<Modification> _modificationsToTarget = null;
    private List<Modification> _modificationsToSource = null;
    private Boolean _childrenNotEqual = null;
    private HashMap<String, List<String>> _allValues = null;

    public SearchResultEntry get_sourceEntry() {
        return _sourceEntry;
    }

    private SearchResultEntry _sourceEntry = null;

    public Entry get_targetEntry() {
        return _targetEntry;
    }

    public void set_targetEntry(SearchResultEntry targetEntry) {
        this._targetEntry = targetEntry;
        if (_dn == null && targetEntry != null) _dn = _targetEntry.getDN();
    }

    private SearchResultEntry _targetEntry = null;


    @Override
    public int compareTo(CompResult o) {
        return getRDN().compareTo(o.getRDN());
    }
    public COMPARE_RESULT get_compare_result() {
        if(_compare_result==null) return COMPARE_RESULT.COMPARE_ERROR;
        return _compare_result;
    }
    public void set_compare_result(COMPARE_RESULT _compare_result) {
        this._compare_result = _compare_result;
    }
    public String getParent_DN() {
        return _parent_DN;
    }
    public void setParent_DN(String parent_DN) {
        this._parent_DN = parent_DN;
    }
    public void set_dn(String _dn) {
        this._dn = _dn;
    }
    public String get_dn() {
        return _dn;
    }
    public String get_rdnDisplay() {
        return _rdnDisplay;
    }
    public void set_rdnDisplay(String _rdnDisplay) {
        this._rdnDisplay = _rdnDisplay;
    }
    public Boolean get_childrenNotEqual() {
        return _childrenNotEqual;
    }
    public void set_childrenNotEqual(Boolean _childrenNotEqual) {
        this._childrenNotEqual = _childrenNotEqual;
    }
    public List<Modification> get_ModificationsToTarget() {
        return _modificationsToTarget;
    }
    public List<Modification> get_ModificationsToSource() {
        return _modificationsToSource;
    }
    public boolean is_entryEqual() {
        return _entryEqual;
    }
    public void set_entryEqual(boolean _entryEqual) {
        this._entryEqual = _entryEqual;
    }
    public ENTRY_TYPE get_entry_type() {
        return _entry_type;
    }
    public void set_entry_type(ENTRY_TYPE _entry_type) {
        this._entry_type = _entry_type;
    }
    public CompResult(SearchResultEntry entry) {
        if (entry != null) _dn = entry.getDN();
        _sourceEntry = entry;
    }
    public List getAttributeValues(String attributeName) {
        return _allValues.get(attributeName);
    }


    public CompResult clone() {
        CompResult compResult = new CompResult(_sourceEntry);
        compResult._targetEntry = _targetEntry;
        compResult._childrenNotEqual = _childrenNotEqual;
        compResult.set_compare_result(get_compare_result());
        compResult._dn = _dn;
        compResult._parent_DN = _parent_DN;
        compResult._entry_type = _entry_type;
        compResult._entryEqual = _entryEqual;
        compResult._rdnDisplay = _rdnDisplay;
        if (_modificationsToTarget != null) {
            compResult._modificationsToTarget = new ArrayList<Modification>();
            compResult._modificationsToTarget.addAll(_modificationsToTarget);
        }
        if (_modificationsToSource != null) {
            compResult._modificationsToSource = new ArrayList<Modification>();
            compResult._modificationsToSource.addAll(_modificationsToSource);
        }
        if (_allValues != null) {
            compResult._allValues = new HashMap<>();
            compResult._allValues.putAll(_allValues);
        }
        return compResult;
    }

    public void addAttributeValue(String attributeName, String value) {
        if (_allValues == null) _allValues = new HashMap<>();
        if (_allValues.containsKey(attributeName)) {
            List values = _allValues.get(attributeName);
            values.add(value);
            _allValues.put(attributeName, values);
        } else {
            List values = new ArrayList();
            values.add(value);
            _allValues.put(attributeName, values);
        }
    }

    private void addAttribute(String attributeName, String[] values) {
        if (values == null || values.length == 0) return;
        for (String v : values) {
            addAttributeValue(attributeName, v);
        }
    }

    public void setAttributes(Collection<Attribute> attributes) {
        for (Attribute at : attributes) {
            addAttribute(at.getName(), at.getValues());
        }
    }

    public List<Modification> findValue(String value) {
        List<Modification> modifications = new ArrayList<>();
        for (String attName : _allValues.keySet()) {
            List<String> values = _allValues.get(attName);
            for (String val : values) {
                List<String> sameValues = new ArrayList<>();
                if (val.contains(value)) {
                    sameValues.add(val);
                }
                if (!sameValues.isEmpty()) {
                    String[] storeValues = new String[sameValues.size()];
                    for (int i = 0; i < sameValues.size(); i++) {
                        storeValues[i] = sameValues.get(i);
                    }
                    Modification mod = new Modification(ModificationType.ADD, attName, storeValues);
                    modifications.add(mod);
                }
            }
        }
        if (modifications.isEmpty()) return null;
        return modifications;
    }

    public void resolveCompareResult(COMPARE_RESULT childResult) {
        if (childResult == null) {
            if (is_entryEqual()) {
                set_compare_result(COMPARE_RESULT.ENTRY_EQUAL);
            } else {
                if (get_entry_type().equals(ENTRY_TYPE.SOURCE)) {
                    set_compare_result(COMPARE_RESULT.ONLY_IN_SOURCE);
                } else if (get_entry_type().equals(ENTRY_TYPE.TARGET)) {
                    set_compare_result(COMPARE_RESULT.ONLY_IN_TARGET);
                } else {
                    set_compare_result(COMPARE_RESULT.ENTRY_NOT_EQUAL);
                }
            }
            return;
        }
        if (_compare_result == null) {
            if (is_entryEqual()) {
                if (childResult.equals(COMPARE_RESULT.ENTRY_EQUAL)) {
                    set_compare_result(COMPARE_RESULT.ENTRY_EQUAL);
                } else {
                    set_compare_result(COMPARE_RESULT.ENTRY_EQUAL_BUT_CHILDREN_NOT);
                }
            } else {
                if (get_entry_type().equals(ENTRY_TYPE.SOURCE)) {
                    set_compare_result(COMPARE_RESULT.ONLY_IN_SOURCE);
                } else if (get_entry_type().equals(ENTRY_TYPE.TARGET)) {
                    set_compare_result(COMPARE_RESULT.ONLY_IN_TARGET);
                } else {
                    set_compare_result(COMPARE_RESULT.ENTRY_NOT_EQUAL);
                }
            }
        } else {
            if (_compare_result.equals(COMPARE_RESULT.ENTRY_EQUAL)) {
                if (!childResult.equals(COMPARE_RESULT.ENTRY_EQUAL)) {
                    set_compare_result(COMPARE_RESULT.ENTRY_EQUAL_BUT_CHILDREN_NOT);
                }
            }
        }
    }

    public void set_ModificationsToTarget(List<Modification> modifications) {
        this._modificationsToTarget = modifications;
        List<Modification> modToSource = new ArrayList<>();
        if (_modificationsToTarget != null && _modificationsToTarget.size() > 0) {
            for (Modification m : _modificationsToTarget) {
                if (m.getModificationType().equals(ModificationType.ADD)) {
                    Modification mNew = new Modification(ModificationType.DELETE, m.getAttributeName(), m.getValues());
                    modToSource.add(mNew);
                } else if (m.getModificationType().equals(ModificationType.DELETE)) {
                    Modification mNew = new Modification(ModificationType.ADD, m.getAttributeName(), m.getValues());
                    modToSource.add(mNew);
                }
            }
        }
        _modificationsToSource = modToSource;
    }

    public List<Modification> get_ModificationsToTarget(String attributeName) {
        List<Modification> returnValue = new ArrayList<>();
        if (_modificationsToTarget != null && _modificationsToTarget.size() > 0) {
            for (Modification m : _modificationsToTarget) {
                if (m.getAttributeName().equalsIgnoreCase(attributeName)) returnValue.add(m);
            }
        }
        return returnValue;
    }

    public List<Modification> get_ModificationsToTarget(List<String> attributes) {
        List<Modification> returnValue = new ArrayList<>();
        HashSet<String> atts = attributes.stream().map(String::toLowerCase).collect(Collectors.toCollection(HashSet::new));

        if (_modificationsToTarget != null && _modificationsToTarget.size() > 0) {
            for (Modification m : _modificationsToTarget) {
                if (atts.contains(m.getAttributeName().toLowerCase())) returnValue.add(m);
            }
        }
        return returnValue;
    }

    public List<Modification> get_ModificationsToSource(List<String> attributes) {
        List<Modification> returnValue = new ArrayList<>();
        HashSet<String> atts = attributes.stream().map(String::toLowerCase).collect(Collectors.toCollection(HashSet::new));
        if (_modificationsToSource != null && _modificationsToSource.size() > 0) {
            for (Modification m : _modificationsToSource) {
                if (atts.contains(m.getAttributeName().toLowerCase())) returnValue.add(m);
            }
        }
        return returnValue;
    }

    public String getRDN() {
        if (_rdnDisplay != null) {
            return _rdnDisplay;
        }
        if (get_dn() != null && !get_dn().equalsIgnoreCase("")) {
            String[] split = get_dn().split(",");
            if (split == null || split.length == 0) {
                return "COULD NOT PARSE RDN 1 ";
            } else {
                String rdnWithAttribute = split[0];
                return rdnWithAttribute;
                //int index = rdnWithAttribute.indexOf("=",0);
                //return rdnWithAttribute.substring(index+1);
            }
        } else {
            return "COULD NOT PARSE RDN 2";
        }
    }

    /**
     * @return Array of differences list.get(0) = attribute name, list.get(1)=ATT_MOD_OPERATION, list.get(2) = old Value, list.get(3) = new value
     */
    public TreeSet<String> getDifferentAttributes() {
        TreeSet<String> diffs = new TreeSet<String>(String.CASE_INSENSITIVE_ORDER);
        if (_modificationsToTarget == null) return diffs;
        for (Modification mod : _modificationsToTarget) {
            String attName = mod.getAttributeName();
            diffs.add(attName);
        }
        return diffs;
    }
    public static boolean stringContains(String sourceValue, String targetValue, boolean exact, boolean matchCase) {
        String src = null;
        String tgt = null;
        if (sourceValue == null) src = "";
        else src = sourceValue;
        if (targetValue == null) tgt = "";
        else tgt = targetValue;
        if (!matchCase) {
            src = src.toLowerCase();
            tgt = tgt.toLowerCase();
        }
        if (!exact) {
            if (src.contains(tgt)) return true;
            else return false;
        } else {
            String pattern = "\\b" + tgt + "\\b";
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(src);
            return m.find();
        }
    }


    public String toString() {
        return getRDN();
    }

    public String print() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("DN=" + _dn + "\n");
        stringBuilder.append("RDN=" + getRDN() + "\n");
        stringBuilder.append("ENTRY TYPE" + _entry_type + "\n");
        stringBuilder.append("ENTRY EQUAL=" + _entryEqual + "\n");
        stringBuilder.append("CHILDREN NOT EQUAL=" + _childrenNotEqual + "\n");
        stringBuilder.append("COMPARE RESULT=" + _compare_result + "\n");

        return stringBuilder.toString();
    }
}

File: src/main/kotlin/ch/vilki/jfxldap/backend/CompResult.kt
package ch.vilki.jfxldap.backend

import com.unboundid.ldap.sdk.*
import java.util.*
import java.util.regex.Pattern
import java.util.stream.Collectors

class KCompResult(entry: SearchResultEntry?) : java.io.Serializable, Comparable<KCompResult>, Cloneable {
    enum class ENTRY_TYPE {
        SOURCE, TARGET, BOTH
    }

    enum class COMPARE_RESULT {
        ENTRY_EQUAL,
        ENTRY_NOT_EQUAL,
        ONLY_IN_SOURCE,
        ONLY_IN_TARGET,
        ENTRY_EQUAL_BUT_CHILDREN_NOT,
        COMPARE_ERROR
    }

    private var _compare_result: COMPARE_RESULT? = null
    private var _dn: String? = null
    private var _parent_DN: String? = null
    private lateinit var _entry_type: ENTRY_TYPE
    private var _entryEqual = false
    private var _rdnDisplay: String? = null
    private var _modificationsToTarget: MutableList<Modification>? = null
    private var _modificationsToSource: MutableList<Modification>? = null
    private var _childrenNotEqual: Boolean? = null
    private var _allValues: HashMap<String, MutableList<String>>? = null
    
    var _sourceEntry: SearchResultEntry? = null
        private set
    
    var _targetEntry: SearchResultEntry? = null
        set(value) {
            field = value
            if (_dn == null && value != null) _dn = value.dn
        }
    
    init {
        if (entry != null) _dn = entry.dn
        _sourceEntry = entry
    }

    override fun compareTo(other: KCompResult): Int {
        return getRDN().compareTo(other.getRDN())
    }

    fun get_compare_result(): COMPARE_RESULT {
        return _compare_result ?: COMPARE_RESULT.COMPARE_ERROR
    }

    fun set_compare_result(compareResult: COMPARE_RESULT) {
        _compare_result = compareResult
    }

    fun getParent_DN(): String? {
        return _parent_DN
    }

    fun setParent_DN(parentDN: String) {
        _parent_DN = parentDN
    }

    fun set_dn(dn: String) {
        _dn = dn
    }

    fun get_dn(): String? {
        return _dn
    }

    fun get_rdnDisplay(): String? {
        return _rdnDisplay
    }

    fun set_rdnDisplay(rdnDisplay: String) {
        _rdnDisplay = rdnDisplay
    }

    fun get_childrenNotEqual(): Boolean? {
        return _childrenNotEqual
    }

    fun set_childrenNotEqual(childrenNotEqual: Boolean?) {
        _childrenNotEqual = childrenNotEqual
    }

    fun get_ModificationsToTarget(): List<Modification>? {
        return _modificationsToTarget
    }

    fun get_ModificationsToSource(): List<Modification>? {
        return _modificationsToSource
    }

    fun is_entryEqual(): Boolean {
        return _entryEqual
    }

    fun set_entryEqual(entryEqual: Boolean) {
        _entryEqual = entryEqual
    }

    fun get_entry_type(): ENTRY_TYPE {
        return _entry_type
    }

    fun set_entry_type(entryType: ENTRY_TYPE) {
        _entry_type = entryType
    }

    fun getAttributeValues(attributeName: String): List<String>? {
        return _allValues?.get(attributeName)
    }

    override fun clone(): KCompResult {
        val compResult = KCompResult(_sourceEntry)
        compResult._targetEntry = _targetEntry
        compResult._childrenNotEqual = _childrenNotEqual
        compResult.set_compare_result(get_compare_result())
        compResult._dn = _dn
        compResult._parent_DN = _parent_DN
        compResult._entry_type = _entry_type
        compResult._entryEqual = _entryEqual
        compResult._rdnDisplay = _rdnDisplay
        
        if (_modificationsToTarget != null) {
            compResult._modificationsToTarget = ArrayList()
            compResult._modificationsToTarget!!.addAll(_modificationsToTarget!!)
        }
        
        if (_modificationsToSource != null) {
            compResult._modificationsToSource = ArrayList()
            compResult._modificationsToSource!!.addAll(_modificationsToSource!!)
        }
        
        if (_allValues != null) {
            compResult._allValues = HashMap()
            compResult._allValues!!.putAll(_allValues!!)
        }
        
        return compResult
    }

    fun addAttributeValue(attributeName: String, value: String) {
        if (_allValues == null) _allValues = HashMap()
        
        if (_allValues!!.containsKey(attributeName)) {
            val values = _allValues!![attributeName]
            values!!.add(value)
            _allValues!![attributeName] = values
        } else {
            val values = ArrayList<String>()
            values.add(value)
            _allValues!![attributeName] = values
        }
    }

    private fun addAttribute(attributeName: String, values: Array<String>?) {
        if (values == null || values.isEmpty()) return
        for (v in values) {
            addAttributeValue(attributeName, v)
        }
    }

    fun setAttributes(attributes: Collection<Attribute>) {
        for (at in attributes) {
            addAttribute(at.name, at.values)
        }
    }

    fun findValue(value: String): List<Modification>? {
        val modifications = ArrayList<Modification>()
        
        _allValues?.let { allValues ->
            for (attName in allValues.keys) {
                val values = allValues[attName]
                values?.let { valuesList ->
                    for (val1 in valuesList) {
                        val sameValues = ArrayList<String>()
                        if (val1.contains(value)) {
                            sameValues.add(val1)
                        }
                        if (sameValues.isNotEmpty()) {
                            val storeValues = sameValues.toTypedArray()
                            val mod = Modification(ModificationType.ADD, attName, *storeValues)
                            modifications.add(mod)
                        }
                    }
                }
            }
        }
        
        return if (modifications.isEmpty()) null else modifications
    }

    fun resolveCompareResult(childResult: COMPARE_RESULT?) {
        if (childResult == null) {
            if (is_entryEqual()) {
                set_compare_result(COMPARE_RESULT.ENTRY_EQUAL)
            } else {
                when {
                    get_entry_type() == ENTRY_TYPE.SOURCE -> set_compare_result(COMPARE_RESULT.ONLY_IN_SOURCE)
                    get_entry_type() == ENTRY_TYPE.TARGET -> set_compare_result(COMPARE_RESULT.ONLY_IN_TARGET)
                    else -> set_compare_result(COMPARE_RESULT.ENTRY_NOT_EQUAL)
                }
            }
            return
        }
        
        if (_compare_result == null) {
            if (is_entryEqual()) {
                if (childResult == COMPARE_RESULT.ENTRY_EQUAL) {
                    set_compare_result(COMPARE_RESULT.ENTRY_EQUAL)
                } else {
                    set_compare_result(COMPARE_RESULT.ENTRY_EQUAL_BUT_CHILDREN_NOT)
                }
            } else {
                when {
                    get_entry_type() == ENTRY_TYPE.SOURCE -> set_compare_result(COMPARE_RESULT.ONLY_IN_SOURCE)
                    get_entry_type() == ENTRY_TYPE.TARGET -> set_compare_result(COMPARE_RESULT.ONLY_IN_TARGET)
                    else -> set_compare_result(COMPARE_RESULT.ENTRY_NOT_EQUAL)
                }
            }
        } else {
            if (_compare_result == COMPARE_RESULT.ENTRY_EQUAL) {
                if (childResult != COMPARE_RESULT.ENTRY_EQUAL) {
                    set_compare_result(COMPARE_RESULT.ENTRY_EQUAL_BUT_CHILDREN_NOT)
                }
            }
        }
    }

    fun set_ModificationsToTarget(modifications: MutableList<Modification>?) {
        _modificationsToTarget = modifications
        val modToSource = ArrayList<Modification>()
        
        _modificationsToTarget?.let { mods ->
            if (mods.isNotEmpty()) {
                for (m in mods) {
                    when (m.modificationType) {
                        ModificationType.ADD -> {
                            val mNew = Modification(ModificationType.DELETE, m.attributeName, *m.values)
                            modToSource.add(mNew)
                        }
                        ModificationType.DELETE -> {
                            val mNew = Modification(ModificationType.ADD, m.attributeName, *m.values)
                            modToSource.add(mNew)
                        }
                        else -> {}
                    }
                }
            }
        }
        
        _modificationsToSource = modToSource
    }

    fun get_ModificationsToTarget(attributeName: String): List<Modification> {
        val returnValue = ArrayList<Modification>()
        
        _modificationsToTarget?.let { mods ->
            if (mods.isNotEmpty()) {
                for (m in mods) {
                    if (m.attributeName.equals(attributeName, ignoreCase = true)) returnValue.add(m)
                }
            }
        }
        
        return returnValue
    }

    fun get_ModificationsToTarget(attributes: List<String>): List<Modification> {
        val returnValue = ArrayList<Modification>()
        val atts = attributes.map { it.lowercase() }.toCollection(HashSet())
        
        _modificationsToTarget?.let { mods ->
            if (mods.isNotEmpty()) {
                for (m in mods) {
                    if (atts.contains(m.attributeName.lowercase())) returnValue.add(m)
                }
            }
        }
        
        return returnValue
    }

    fun get_ModificationsToSource(attributes: List<String>): List<Modification> {
        val returnValue = ArrayList<Modification>()
        val atts = attributes.map { it.lowercase() }.toCollection(HashSet())
        
        _modificationsToSource?.let { mods ->
            if (mods.isNotEmpty()) {
                for (m in mods) {
                    if (atts.contains(m.attributeName.lowercase())) returnValue.add(m)
                }
            }
        }
        
        return returnValue
    }

    fun getRDN(): String {
        _rdnDisplay?.let {
            return it
        }
        
        if (!get_dn().isNullOrEmpty()) {
            val split = get_dn()!!.split(",").toTypedArray()
            return if (split.isEmpty()) {
                "COULD NOT PARSE RDN 1 "
            } else {
                split[0]
            }
        } else {
            return "COULD NOT PARSE RDN 2"
        }
    }

    /**
     * @return Set of attribute names that have differences
     */
    fun getDifferentAttributes(): TreeSet<String> {
        val diffs = TreeSet<String>(String.CASE_INSENSITIVE_ORDER)
        
        _modificationsToTarget?.let { mods ->
            for (mod in mods) {
                val attName = mod.attributeName
                diffs.add(attName)
            }
        }
        
        return diffs
    }

    companion object {
        fun stringContains(sourceValue: String?, targetValue: String?, exact: Boolean, matchCase: Boolean): Boolean {
            val src = sourceValue ?: ""
            val tgt = targetValue ?: ""
            
            val srcFinal = if (!matchCase) src.lowercase() else src
            val tgtFinal = if (!matchCase) tgt.lowercase() else tgt
            
            if (!exact) {
                return srcFinal.contains(tgtFinal)
            } else {
                val pattern = "\\b$tgtFinal\\b"
                val p = Pattern.compile(pattern)
                val m = p.matcher(srcFinal)
                return m.find()
            }
        }
    }

    override fun toString(): String {
        return getRDN()
    }

    fun print(): String {
        val stringBuilder = StringBuilder()
        stringBuilder.append("DN=$_dn\n")
        stringBuilder.append("RDN=${getRDN()}\n")
        stringBuilder.append("ENTRY TYPE$_entry_type\n")
        stringBuilder.append("ENTRY EQUAL=$_entryEqual\n")
        stringBuilder.append("CHILDREN NOT EQUAL=$_childrenNotEqual\n")
        stringBuilder.append("COMPARE RESULT=$_compare_result\n")
        
        return stringBuilder.toString()
    }
}
